File: python/kohaku/src/kohaku/evals/dataset.py
"""Distillation dataset export (port of TS packages/evals/src/dataset.ts).

Research on catalog-constrained declarative UI generation shows small models close much of the gap to
frontier models when trained/prompted on this kind of constrained output. A human-approved fixation
(L0, ``FixationRecord.pinnedSpec``) is the best teacher example available: it is exactly the
``{components, events}`` pair a distilled model should learn to reproduce for a given Intent, already
vetted by a human approver. Golden Specs (kohaku.evals's own regression fixtures) are the same shape of
teacher example without the human-approval step, useful as supplementary training data.

Exported as JSONL (one line per Spec) so it composes with the usual fine-tuning / distillation tooling
(streaming read, ``wc -l``, ``shuf``, ...) without a custom parser.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from kohaku.spec import ComponentNode, FixationRecord, Intent, UISpec, canonical_stringify

from .judge import ColumnMeta


@dataclass(frozen=True)
class DistillationRecord:
    """One distillation-dataset record: the LLM's declarative-generation input/output pair."""

    intent: dict[str, Any]
    """{"canonical": str, "params": dict[str, JsonValue]}."""
    refs: list[str]
    """The set of $refs the Spec's components read from (deduplicated, first-occurrence order)."""
    target: dict[str, Any]
    """The LLM's actual output range only: {"components": [...], "events": [...]}. Deliberately excludes
    kohaku (the protocol-envelope version), provenance, and dataVersion -- composer fills all three in
    around the model's output, so a distillation target that included them would teach a distilled model
    to imitate infra plumbing it never actually produces, rather than the declarative UI itself."""
    source: str
    """"fixation" | "golden"."""
    meta: dict[str, Any]
    """{"fixatedAt"?: str, "structureHash"?: str, "tenant"?: str, "catalogFingerprint"?: str}: all four keys
    are populated from the FixationRecord for source="fixation" (each omitted, not written as null, when the
    record predates that field -- Python's canonical_stringify writes None as JSON null rather than
    dropping the key the way TS's does with undefined, so the key must be left out explicitly to stay
    byte-compatible). Empty for source="golden" (golden Specs carry no tenant or fixation-time fingerprint)."""
    shape: list[ColumnMeta] | None = None
    """Optional column metadata for the Intent's data shape (from describe_shape; omitted when None)."""


def _refs_of(components: list[ComponentNode]) -> list[str]:
    """Extracts the set of $refs a Spec's components read from (deduplicated, first-occurrence order)."""
    seen: set[str] = set()
    out: list[str] = []
    for c in components:
        if c.data is not None and c.data.ref not in seen:
            seen.add(c.data.ref)
            out.append(c.data.ref)
    return out


def _record_to_json(record: DistillationRecord) -> dict[str, Any]:
    out: dict[str, Any] = {
        "intent": record.intent,
        "refs": record.refs,
        "target": record.target,
        "source": record.source,
        "meta": record.meta,
    }
    if record.shape is not None:
        out["shape"] = [
            {k: v for k, v in (("name", s.name), ("type", s.type), ("description", s.description)) if v is not None}
            for s in record.shape
        ]
    return out


def _fixation_meta(record: FixationRecord) -> dict[str, Any]:
    meta: dict[str, Any] = {"fixatedAt": record.fixatedAt, "structureHash": record.structureHash}
    if record.tenant is not None:
        meta["tenant"] = record.tenant
    if record.catalogFingerprint is not None:
        meta["catalogFingerprint"] = record.catalogFingerprint
    return meta


def _source_order(source: str) -> int:
    """Explicit tie-break order for entries that share the same sort key (see the sort below)."""
    return 0 if source == "fixation" else 1


def export_distillation_dataset(
    fixations: list[FixationRecord],
    golden: list[UISpec] | None = None,
    *,
    describe_shape: Callable[[Intent], list[ColumnMeta] | None] | None = None,
    tenant: str | None = None,
) -> str:
    """Exports human-approved fixations (and optionally golden Specs) as a JSONL distillation dataset: one
    canonical-JSON line per Spec.

    ``tenant``: restricts the export to fixations whose ``tenant`` matches this id (golden Specs carry no
    tenant and are always included regardless of this filter). ``None`` (the default) includes every tenant
    present in ``fixations`` in one JSONL -- the on-disk fixations snapshot legitimately holds records for
    several tenants side by side (keyed ``f"{tenant} {intentHash}"``; see storage-port.ts), so without this
    filter their approved structures land unmarked in one file unless ``meta["tenant"]`` is inspected per
    line.

    Deterministic order: entries are sorted by ``(intentHash, source)`` ascending (``FixationRecord.intentHash``
    for fixation entries, ``spec.intent.hash`` for golden entries; a "fixation" entry sorts before a "golden"
    entry that shares the same intentHash -- an explicit tie-break rather than relying on sort stability),
    and each line is serialized with canonical_stringify (deep key-sorted, byte-identical across
    languages). Re-running the export on the same input therefore reproduces byte-identical output -- this
    is also what pins the TS/Python cross-language golden (spec/scripts/generate-cross-language-fixtures.ts).

    Returns the empty string for empty input (no trailing newline); otherwise every line, including the
    last, ends with "\\n".
    """
    keyed: list[tuple[str, DistillationRecord]] = []

    selected_fixations = [r for r in fixations if r.tenant == tenant] if tenant is not None else fixations

    for record in selected_fixations:
        spec = record.pinnedSpec
        shape = describe_shape(spec.intent) if describe_shape is not None else None
        keyed.append(
            (
                record.intentHash,
                DistillationRecord(
                    intent={"canonical": spec.intent.canonical, "params": spec.intent.params},
                    refs=_refs_of(spec.components),
                    shape=shape,
                    target={
                        "components": [c.to_wire() for c in spec.components],
                        "events": [e.to_wire() for e in spec.events],
                    },
                    source="fixation",
                    meta=_fixation_meta(record),
                ),
            )
        )

    for spec in golden or []:
        shape = describe_shape(spec.intent) if describe_shape is not None else None
        keyed.append(
            (
                spec.intent.hash,
                DistillationRecord(
                    intent={"canonical": spec.intent.canonical, "params": spec.intent.params},
                    refs=_refs_of(spec.components),
                    shape=shape,
                    target={
                        "components": [c.to_wire() for c in spec.components],
                        "events": [e.to_wire() for e in spec.events],
                    },
                    source="golden",
                    meta={},
                ),
            )
        )

    keyed.sort(key=lambda kv: (kv[0], _source_order(kv[1].source)))
    lines = [canonical_stringify(_record_to_json(record)) for _, record in keyed]
    return "".join(f"{line}\n" for line in lines)

File: packages/evals/src/dataset.ts
import {
  type CanonicalIntent,
  type ComponentNode,
  canonicalStringify,
  type EventBinding,
  type FixationRecord,
  type UISpec,
} from "@kohaku-ui/spec-core";
import type { ColumnMeta } from "./judge.js";

/**
 * Distillation dataset.
 *
 * Research on catalog-constrained declarative UI generation shows small models close much of the gap to
 * frontier models when trained/prompted on this kind of constrained output. A human-approved fixation
 * (L0, `FixationRecord.pinnedSpec`) is the best teacher example available: it is exactly the
 * {components, events} pair a distilled model should learn to reproduce for a given Intent, already
 * vetted by a human approver. Golden Specs (packages/evals's own regression fixtures) are the same shape
 * of teacher example without the human-approval step, useful as supplementary training data.
 *
 * Exported as JSONL (one line per Spec) so it composes with the usual fine-tuning / distillation tooling
 * (streaming read, `wc -l`, `shuf`, …) without a custom parser.
 */

/** One distillation-dataset record: the LLM's declarative-generation input/output pair. */
export interface DistillationRecord {
  intent: { canonical: string; params: Record<string, unknown> };
  /** The set of $refs the Spec's components read from (deduplicated, first-occurrence order). */
  refs: string[];
  /** Optional column metadata for the Intent's data shape (from `describeShape`; omitted when not supplied). */
  shape?: ColumnMeta[];
  /**
   * The LLM's actual output range only: `components` + `events`. Deliberately excludes `kohaku` (the
   * protocol-envelope version), `provenance`, and `dataVersion` — composer fills all three in around the
   * model's output, so a distillation target that included them would teach a distilled model to imitate
   * infra plumbing it never actually produces, rather than the declarative UI itself.
   */
  target: { components: ComponentNode[]; events: EventBinding[] };
  source: "fixation" | "golden";
  /**
   * `tenant` / `catalogFingerprint` are populated from the FixationRecord for `source: "fixation"` rows
   * (omitted when the record predates that field, exactly like `FixationRecordSchema`'s own optionality).
   * Golden-derived rows keep an empty `meta` (golden Specs carry no tenant or fixation-time fingerprint).
   * Without `tenant`, a dataset spanning several tenants (the on-disk `fixations.json` legitimately holds
   * one record per (tenant, intentHash), see storage-port.ts) cannot be told apart after export; without
   * `catalogFingerprint`, there is no record of which catalog generation `target`'s component
   * types/versions were valid for.
   */
  meta: { fixatedAt?: string; structureHash?: string; tenant?: string; catalogFingerprint?: string };
}

export interface ExportDistillationDatasetInput {
  /** Human-approved L1→L0 fixations (validated FixationRecordSchema data; see @kohaku-ui/spec-core). */
  fixations: FixationRecord[];
  /** Optional supplementary teacher examples (e.g. golden regression fixtures' `expected` Specs). */
  golden?: UISpec[];
}

export interface ExportDistillationDatasetOptions {
  /**
   * Optional per-record column-metadata lookup (typically backed by the domain's resultShape). Return
   * undefined to omit the `shape` field for that record.
   */
  describeShape?: (intent: CanonicalIntent) => ColumnMeta[] | undefined;
  /**
   * Restrict the export to fixations whose `tenant` matches this id (`golden` Specs carry no tenant and
   * are always included regardless of this filter). Omitted = every tenant present in `input.fixations`
   * is included in one JSONL — the on-disk `fixations.json` snapshot legitimately holds records for
   * several tenants side by side (keyed `${tenant} ${intentHash}`; see storage-port.ts), so without this
   * filter their approved structures land unmarked in one file unless `meta.tenant` is inspected per line.
   */
  tenant?: string;
}

/** Extracts the set of $refs a Spec's components read from (deduplicated, first-occurrence order). */
function refsOf(components: ComponentNode[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const c of components) {
    const ref = c.data?.$ref;
    if (ref != null && !seen.has(ref)) {
      seen.add(ref);
      out.push(ref);
    }
  }
  return out;
}

/**
 * Projects a ColumnMeta down to exactly `{name, type?, description?}`. `ColumnMeta` is structurally typed
 * (callers commonly pass an object shaped by the domain's own resultShape type), so without this
 * projection any extra keys that object happens to carry would leak into the dataset verbatim, breaking
 * the claim that TypeScript's and Python's `.data` (which already projects to these three keys) produce
 * the same dataset.
 */
function projectColumnMeta(shape: ColumnMeta[]): ColumnMeta[] {
  return shape.map((c) => ({
    name: c.name,
    ...(c.type !== undefined ? { type: c.type } : {}),
    ...(c.description !== undefined ? { description: c.description } : {}),
  }));
}

function shapeField(shape: ColumnMeta[] | undefined): { shape: ColumnMeta[] } | Record<string, never> {
  return shape != null ? { shape: projectColumnMeta(shape) } : {};
}

/** Explicit tie-break order for entries that share the same sort key (see the sort below). */
function sourceOrder(source: DistillationRecord["source"]): 0 | 1 {
  return source === "fixation" ? 0 : 1;
}

/**
 * Exports human-approved fixations (and optionally golden Specs) as a JSONL distillation dataset: one
 * canonical-JSON line per Spec.
 *
 * Deterministic order: entries are sorted by `(intentHash, source)` ascending (`FixationRecord.intentHash`
 * for fixation entries, `spec.intent.hash` for golden entries; a `fixation` entry sorts before a `golden`
 * entry that shares the same intentHash — an explicit tie-break rather than relying on sort stability),
 * and each line is serialized with `canonicalStringify` (deep key-sorted, byte-identical across
 * languages). Re-running the export on the same input therefore reproduces byte-identical output — this
 * is also what pins the TS/Python cross-language golden (spec/scripts/generate-cross-language-fixtures.ts).
 *
 * Returns the empty string for an empty input (no trailing newline); otherwise every line, including the
 * last, ends with `\n`.
 */
export function exportDistillationDataset(
  input: ExportDistillationDatasetInput,
  options: ExportDistillationDatasetOptions = {},
): string {
  const keyed: { key: string; record: DistillationRecord }[] = [];

  const fixations =
    options.tenant != null ? input.fixations.filter((r) => r.tenant === options.tenant) : input.fixations;

  for (const record of fixations) {
    const spec = record.pinnedSpec;
    keyed.push({
      key: record.intentHash,
      record: {
        intent: { canonical: spec.intent.canonical, params: spec.intent.params },
        refs: refsOf(spec.components),
        ...shapeField(options.describeShape?.(spec.intent)),
        target: { components: spec.components, events: spec.events },
        source: "fixation",
        meta: {
          fixatedAt: record.fixatedAt,
          structureHash: record.structureHash,
          tenant: record.tenant,
          catalogFingerprint: record.catalogFingerprint,
        },
      },
    });
  }

  for (const spec of input.golden ?? []) {
    keyed.push({
      key: spec.intent.hash,
      record: {
        intent: { canonical: spec.intent.canonical, params: spec.intent.params },
        refs: refsOf(spec.components),
        ...shapeField(options.describeShape?.(spec.intent)),
        target: { components: spec.components, events: spec.events },
        source: "golden",
        meta: {},
      },
    });
  }

  keyed.sort((a, b) => {
    if (a.key !== b.key) return a.key < b.key ? -1 : 1;
    return sourceOrder(a.record.source) - sourceOrder(b.record.source);
  });
  const lines = keyed.map(({ record }) => canonicalStringify(record));
  return lines.length > 0 ? `${lines.join("\n")}\n` : "";
}
